File: documentation/api/examples/typescript/websocket_events.ts
/**
 * Example: Real-time event streaming via WebSocket
 *
 * This example demonstrates how to connect to the WebSocket endpoint
 * and receive real-time events about workflows, executions, and work items.
 *
 * Install: npm install ws @types/ws
 */
import WebSocket from "ws";

// Configuration
const WS_URL = "ws://localhost:8000/api/v2/events/stream";
const API_TOKEN = "your_token_here"; // Get from server startup logs

// Type definitions
interface CodetoreumEvent {
  type: string;
  timestamp: string;
  data: Record<string, any>;
}

type EventHandler = (event: CodetoreumEvent) => void | Promise<void>;

/**
 * WebSocket client for Codetoreum event streaming
 */
class CodetoreumEventStream {
  private ws: WebSocket | null = null;
  private handlers: Map<string, EventHandler[]> = new Map();
  private url: string;

  constructor(token: string, baseUrl: string = WS_URL) {
    this.url = `${baseUrl}?token=${token}`;
  }

  /**
   * Connect to the WebSocket
   */
  async connect(): Promise<void> {
    return new Promise((resolve, reject) => {
      console.log(`Connecting to ${this.url}...`);

      this.ws = new WebSocket(this.url);

      this.ws.on("open", () => {
        console.log("✓ Connected to event stream");
        resolve();
      });

      this.ws.on("error", (error) => {
        console.error("WebSocket error:", error);
        reject(error);
      });

      this.ws.on("close", () => {
        console.log("✓ Disconnected from event stream");
      });

      this.ws.on("message", (data: WebSocket.Data) => {
        this.handleMessage(data.toString());
      });
    });
  }

  /**
   * Disconnect from the WebSocket
   */
  disconnect(): void {
    if (this.ws) {
      this.ws.close();
      this.ws = null;
    }
  }

  /**
   * Register event handler for specific event type
   */
  on(eventType: string, handler: EventHandler): void {
    if (!this.handlers.has(eventType)) {
      this.handlers.set(eventType, []);
    }
    this.handlers.get(eventType)!.push(handler);
  }

  /**
   * Register handler for all events
   */
  onAll(handler: EventHandler): void {
    this.on("*", handler);
  }

  /**
   * Handle incoming message
   */
  private async handleMessage(message: string): Promise<void> {
    try {
      const event: CodetoreumEvent = JSON.parse(message);
      const eventType = event.type || "unknown";

      const timestamp = new Date().toLocaleTimeString();
      console.log(`[${timestamp}] Received: ${eventType}`);

      // Call specific handlers
      const specificHandlers = this.handlers.get(eventType) || [];
      for (const handler of specificHandlers) {
        await handler(event);
      }

      // Call wildcard handlers
      const wildcardHandlers = this.handlers.get("*") || [];
      for (const handler of wildcardHandlers) {
        await handler(event);
      }
    } catch (error) {
      console.error("Error handling message:", error);
    }
  }

  /**
   * Listen for events (keeps connection open)
   */
  async listen(): Promise<void> {
    if (!this.ws) {
      throw new Error("Not connected. Call connect() first.");
    }

    return new Promise((resolve, reject) => {
      if (!this.ws) {
        reject(new Error("WebSocket not connected"));
        return;
      }

      this.ws.on("close", () => resolve());
      this.ws.on("error", reject);
    });
  }
}

// Example event handlers

async function handleWorkflowStarted(event: CodetoreumEvent): Promise<void> {
  const data = event.data;
  console.log(`  🚀 Workflow started:`);
  console.log(`     Run ID: ${data.workflow_run_id}`);
  console.log(`     Work Item: ${data.work_item_id}`);
  console.log(`     Workflow: ${data.workflow_name}`);
}

async function handleWorkflowCompleted(event: CodetoreumEvent): Promise<void> {
  const data = event.data;
  console.log(`  ✓ Workflow completed:`);
  console.log(`     Run ID: ${data.workflow_run_id}`);
  console.log(`     Status: ${data.status}`);
  console.log(`     Duration: ${data.duration_seconds}s`);
}

async function handleExecutionStarted(event: CodetoreumEvent): Promise<void> {
  const data = event.data;
  console.log(`  ▶ Execution started:`);
  console.log(`     Execution ID: ${data.execution_id}`);
  console.log(`     Agent: ${data.agent_id}`);
  console.log(`     Stage: ${data.stage_name}`);
}

async function handleExecutionProgress(event: CodetoreumEvent): Promise<void> {
  const data = event.data;
  const progress = data.progress || {};
  const pct = progress.percentage || 0;
  const step = progress.current_step || "";
  console.log(`  ⏳ Execution progress: ${pct}% - ${step}`);
}

async function handleExecutionCompleted(event: CodetoreumEvent): Promise<void> {
  const data = event.data;
  console.log(`  ✓ Execution completed:`);
  console.log(`     Execution ID: ${data.execution_id}`);
  console.log(`     Status: ${data.status}`);
}

async function handleExecutionFailed(event: CodetoreumEvent): Promise<void> {
  const data = event.data;
  console.log(`  ✗ Execution failed:`);
  console.log(`     Execution ID: ${data.execution_id}`);
  console.log(`     Error: ${data.error_message}`);
}

async function handleWorkItemUpdated(event: CodetoreumEvent): Promise<void> {
  const data = event.data;
  console.log(`  📝 Work item updated:`);
  console.log(`     Work Item ID: ${data.work_item_id}`);
  console.log(`     Status: ${data.status}`);
  console.log(`     Stage: ${data.workflow_stage}`);
}

async function logAllEvents(event: CodetoreumEvent): Promise<void> {
  // Pretty print the entire event (for debugging)
  console.log(`  Raw event:`, JSON.stringify(event, null, 2));
}

/**
 * Main example: Monitor all workflow and execution events
 */
async function main() {
  const client = new CodetoreumEventStream(API_TOKEN);

  // Register event handlers
  client.on("workflow.started", handleWorkflowStarted);
  client.on("workflow.completed", handleWorkflowCompleted);
  client.on("execution.started", handleExecutionStarted);
  client.on("execution.progress", handleExecutionProgress);
  client.on("execution.completed", handleExecutionCompleted);
  client.on("execution.failed", handleExecutionFailed);
  client.on("work_item.updated", handleWorkItemUpdated);

  // Optionally log all events for debugging
  // client.onAll(logAllEvents);

  // Connect and listen
  await client.connect();

  console.log("\n📡 Listening for events... (Press Ctrl+C to stop)\n");

  // Handle graceful shutdown
  process.on("SIGINT", () => {
    console.log("\n\nStopping...");
    client.disconnect();
    process.exit(0);
  });

  // Listen for events (this keeps the process running)
  try {
    await client.listen();
  } catch (error) {
    console.error("Error:", error);
    client.disconnect();
    process.exit(1);
  }
}

/**
 * Example: Monitor only specific execution
 */
async function exampleFilteredMonitoring() {
  const client = new CodetoreumEventStream(API_TOKEN);

  // Track specific execution
  const targetExecutionId = "exec_abc123";

  const handleTargetExecution = async (event: CodetoreumEvent) => {
    const data = event.data;
    if (data.execution_id === targetExecutionId) {
      console.log(`Target execution event: ${event.type}`);
      console.log(`  Data:`, JSON.stringify(data, null, 2));
    }
  };

  // Register handler for all execution events
  client.on("execution.started", handleTargetExecution);
  client.on("execution.progress", handleTargetExecution);
  client.on("execution.completed", handleTargetExecution);
  client.on("execution.failed", handleTargetExecution);

  await client.connect();

  try {
    await client.listen();
  } finally {
    client.disconnect();
  }
}

// Run if executed directly
if (require.main === module) {
  // Run main example
  main();

  // Or run filtered monitoring example
  // exampleFilteredMonitoring();
}

export { CodetoreumEventStream };
export type { CodetoreumEvent, EventHandler };

File: documentation/api/examples/python/websocket_events.py
"""
Example: Real-time event streaming via WebSocket

This example demonstrates how to connect to the WebSocket endpoint
and receive real-time events about workflows, executions, and work items.
"""
import asyncio
import json
from typing import Callable, Optional
import websockets
from datetime import datetime


# Configuration
WS_URL = "ws://localhost:8000/api/v2/events/stream"
API_TOKEN = "your_token_here"  # Get from server startup logs


class CodetoreumEventStream:
    """WebSocket client for Codetoreum event streaming."""

    def __init__(self, token: str, base_url: str = WS_URL):
        """
        Initialize event stream client.

        Args:
            token: Authentication token
            base_url: WebSocket URL (default: ws://localhost:8000/api/v2/events/stream)
        """
        self.token = token
        self.base_url = f"{base_url}?token={token}"
        self.websocket: Optional[websockets.WebSocketClientProtocol] = None
        self.handlers = {}

    async def connect(self):
        """Connect to the WebSocket."""
        print(f"Connecting to {self.base_url}...")
        self.websocket = await websockets.connect(self.base_url)
        print("✓ Connected to event stream")

    async def disconnect(self):
        """Disconnect from the WebSocket."""
        if self.websocket:
            await self.websocket.close()
            print("✓ Disconnected from event stream")

    def on(self, event_type: str, handler: Callable):
        """
        Register event handler.

        Args:
            event_type: Event type to handle (e.g., "workflow.started")
            handler: Async function to call when event is received
        """
        if event_type not in self.handlers:
            self.handlers[event_type] = []
        self.handlers[event_type].append(handler)

    def on_all(self, handler: Callable):
        """
        Register handler for all events.

        Args:
            handler: Async function to call for all events
        """
        self.on("*", handler)

    async def listen(self):
        """
        Listen for events and dispatch to handlers.

        This is a blocking call that will run until the connection is closed.
        """
        if not self.websocket:
            raise RuntimeError("Not connected. Call connect() first.")

        try:
            async for message in self.websocket:
                # Parse event
                event = json.loads(message)
                event_type = event.get("type", "unknown")

                timestamp = datetime.now().strftime("%H:%M:%S")
                print(f"[{timestamp}] Received: {event_type}")

                # Call specific handlers
                if event_type in self.handlers:
                    for handler in self.handlers[event_type]:
                        await handler(event)

                # Call wildcard handlers
                if "*" in self.handlers:
                    for handler in self.handlers["*"]:
                        await handler(event)

        except websockets.exceptions.ConnectionClosed as e:
            print(f"Connection closed: {e.code} - {e.reason}")
        except websockets.exceptions.WebSocketException as e:
            print(f"WebSocket error: {e}")
            raise
        except json.JSONDecodeError as e:
            print(f"Failed to parse event JSON: {e}")
            raise
        except KeyError as e:
            print(f"Missing expected field in event: {e}")
            raise
        except Exception as e:
            print(f"Unexpected error: {type(e).__name__}: {e}")
            raise


# Example event handlers

async def handle_workflow_started(event):
    """Handle workflow.started events."""
    data = event.get("data", {})
    print(f"  🚀 Workflow started:")
    print(f"     Run ID: {data.get('workflow_run_id')}")
    print(f"     Work Item: {data.get('work_item_id')}")
    print(f"     Workflow: {data.get('workflow_name')}")


async def handle_workflow_completed(event):
    """Handle workflow.completed events."""
    data = event.get("data", {})
    print(f"  ✓ Workflow completed:")
    print(f"     Run ID: {data.get('workflow_run_id')}")
    print(f"     Status: {data.get('status')}")
    print(f"     Duration: {data.get('duration_seconds')}s")


async def handle_execution_started(event):
    """Handle execution.started events."""
    data = event.get("data", {})
    print(f"  ▶ Execution started:")
    print(f"     Execution ID: {data.get('execution_id')}")
    print(f"     Agent: {data.get('agent_id')}")
    print(f"     Stage: {data.get('stage_name')}")


async def handle_execution_progress(event):
    """Handle execution.progress events."""
    data = event.get("data", {})
    progress = data.get("progress", {})
    pct = progress.get("percentage", 0)
    step = progress.get("current_step", "")
    print(f"  ⏳ Execution progress: {pct}% - {step}")


async def handle_execution_completed(event):
    """Handle execution.completed events."""
    data = event.get("data", {})
    print(f"  ✓ Execution completed:")
    print(f"     Execution ID: {data.get('execution_id')}")
    print(f"     Status: {data.get('status')}")


async def handle_execution_failed(event):
    """Handle execution.failed events."""
    data = event.get("data", {})
    print(f"  ✗ Execution failed:")
    print(f"     Execution ID: {data.get('execution_id')}")
    print(f"     Error: {data.get('error_message')}")


async def handle_work_item_updated(event):
    """Handle work_item.updated events."""
    data = event.get("data", {})
    print(f"  📝 Work item updated:")
    print(f"     Work Item ID: {data.get('work_item_id')}")
    print(f"     Status: {data.get('status')}")
    print(f"     Stage: {data.get('workflow_stage')}")


async def log_all_events(event):
    """Log all events (for debugging)."""
    # Pretty print the entire event
    print(f"  Raw event: {json.dumps(event, indent=2)}")


async def main():
    """Example usage: Monitor all workflow and execution events."""
    client = CodetoreumEventStream(token=API_TOKEN)

    # Register event handlers
    client.on("workflow.started", handle_workflow_started)
    client.on("workflow.completed", handle_workflow_completed)
    client.on("execution.started", handle_execution_started)
    client.on("execution.progress", handle_execution_progress)
    client.on("execution.completed", handle_execution_completed)
    client.on("execution.failed", handle_execution_failed)
    client.on("work_item.updated", handle_work_item_updated)

    # Optionally log all events for debugging
    # client.on_all(log_all_events)

    # Connect and listen
    await client.connect()

    try:
        print("\n📡 Listening for events... (Press Ctrl+C to stop)\n")
        await client.listen()
    except KeyboardInterrupt:
        print("\n\nStopping...")
    finally:
        await client.disconnect()


async def example_filtered_monitoring():
    """Example: Monitor only specific execution."""
    client = CodetoreumEventStream(token=API_TOKEN)

    # Track specific execution
    target_execution_id = "exec_abc123"

    async def handle_target_execution(event):
        data = event.get("data", {})
        if data.get("execution_id") == target_execution_id:
            print(f"Target execution event: {event.get('type')}")
            print(f"  Data: {json.dumps(data, indent=2)}")

    # Register handler for all execution events
    client.on("execution.started", handle_target_execution)
    client.on("execution.progress", handle_target_execution)
    client.on("execution.completed", handle_target_execution)
    client.on("execution.failed", handle_target_execution)

    await client.connect()
    try:
        await client.listen()
    finally:
        await client.disconnect()


if __name__ == "__main__":
    # Run main example
    asyncio.run(main())

    # Or run filtered monitoring example
    # asyncio.run(example_filtered_monitoring())
